File: src/greedy/조이스틱.java
package greedy;

import java.util.*;

public class 조이스틱 {

    public int solution(String name) {
        int answer = 0;
        int len = name.length();
        int min = len-1;
        for(int i = 0 ; i < len ; i++){
            char target = name.charAt(i);
            if(target != 'A'){
                answer += Math.min(target-'A','Z'- target + 1);
                int next = i+1;
                while(next < len && name.charAt(next) == 'A') next++;
                min = Math.min(min, (i * 2) + len - next);
            }
        }
        answer += min;
        return answer;
    }
}

File: 그리디/조이스틱.py
def solution(name):
    answer, i = 0, 0
    upDown = []
    for n in name:
        upDown.append(min(ord(n)-ord('A'), ord('Z')-ord(n)+1))
    
    while True:
        answer += upDown[i]
        upDown[i] = 0
        if sum(upDown) == 0:
            break
        
        left, right = 1, 1
        while upDown[i-left] == 0:
            left += 1
        while upDown[i+right] == 0:
            right += 1
        
        if left < right:
            i -= left
            answer += left
        else:
            i += right
            answer += right
            
    return answer
